### PR TITLE
Add PHPUnit tests workflow for multisite

### LINKED ISSUE #5701: Add PHPUnit test workflow for multisite
## Feature description

Some relatively recent additions to the plugin have involved small customizations related to multisite environments ([example](https://github.com/ampproject/amp-wp/blob/5ee474c9de85de8a15494669f92e926b45755ddb/src/BackgroundTask/CronBasedBackgroundTask.php#L103)). Currently we only run PHPUnit tests in single-site environments, so there's no easy way to write tests for these features. 

Potentially useful for reference: 

https://github.com/WordPress/wordpress-develop/blob/master/tests/phpunit/multisite.xml
https://github.com/WordPress/wordpress-develop/blob/master/.github/workflows/phpunit-tests.yml#L257

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by thelovekesh**
## Summary

Fixes #5701 

This PR also updates deprecated `set-output` commands in GitHub Actions workflow files. See https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by thelovekesh** [comment]
@westonruter I have configured workflow for both testsuites i.e. `default` and `external-http`. Some tests seem to be failing mainly from helper functions.  Should I fix them in the same PR? Can you please take a look if any test is need to be skipped?

**Comment by westonruter** [comment]
Interesting. Yes, they should be fixed in this PR. For each failure, I'd suggest checking to see what the cause of the failure is. It may be due to a regular admin user not having all the same privileges as a superadmin on multisite. Therefore the fix may be just to check if its multisite, and then grant the user being tested with superadmin privileges.

Some tests probably also should be skipped if the functionality specifically is not available in multisite. For example, installing themes from the onboarding wizard I don't think is available on multisite.

It'll be interesting to see if this uncovers actual bugs with the code and not just incomplete tests!

**Comment by thelovekesh** [comment]
@westonruter I believe the capability here should be `install_themes`. Thoughts?

https://github.com/ampproject/amp-wp/blob/650a4eb1a693c51a638527f7864ee17161762723/src/Admin/AfterActivationSiteScan.php#L214

**Comment by thelovekesh** [comment]
@westonruter Same issue as https://github.com/ampproject/amp-wp/pull/7288#issuecomment-1321774813 is also here but strangely it's not happening in multisite tests(using wp-env).

**Comment by westonruter** [comment]
> Same issue as [#7288 (comment)](https://github.com/ampproject/amp-wp/pull/7288#issuecomment-1321774813) is also here but strangely it's not happening in multisite tests(using wp-env).

@thelovekesh That's strange. I don't know what's going on. Perhaps just cherry-pick my commit as described in https://github.com/ampproject/amp-wp/pull/7288#issuecomment-1322614600.

File: .github/workflows/build-test-measure.yml [hunk 1]
```diff
@@ -569,6 +569,61 @@ jobs:
 
   #-----------------------------------------------------------------------------------------------------------------------
 
+  unit-test-multisite-php:
+    name: "Unit Tests Multisite: PHP 7.4, WP Latest"
+    needs: pre-run
+    runs-on: ubuntu-latest
+    # if: needs.pre-run.outputs.changed-php-count > 0
```
Comment (westonruter) [inline .github/workflows/build-test-measure.yml]: Is this comment needed?

File: tests/php/src/Admin/AfterActivationSiteScanTest.php [hunk 2]
```diff
@@ -276,7 +276,18 @@ function ( $caps, $cap ) {
...
+				[
+					'role' => 'administrator',
+				]
+			);
+
+			grant_super_admin( $user_id );
+			wp_set_current_user( $user_id );
+		} else {
+			wp_set_current_user( self::factory()->user->create( [ 'role' => 'administrator' ] ) );
+		}
```
Comment (westonruter) [inline tests/php/src/Admin/AfterActivationSiteScanTest.php]: Since this same logic is being repeated in multiple files, I suggest factoring it out into a [shared trait](https://github.com/ampproject/amp-wp/tree/develop/tests/php/src/Helpers) or else to put in the base [`TestCase`](https://github.com/ampproject/amp-wp/blob/develop/tests/php/src/TestCase.php).

File: tests/php/src/ReaderThemeLoaderTest.php [hunk 3]
```diff
@@ -151,12 +151,29 @@ public function test_filter_wp_prepare_themes_to_indicate_reader_theme() {
...
+		}
 
 		// Note that this is added via filter and not called directly because the filtered value is keyed by theme slug,
 		// but the return value of wp_prepare_themes_for_js() is keyed with numeric indices.
 		$this->instance->register(); // This adds a `wp_prepare_themes_for_js` filter.
-		$themes = wp_prepare_themes_for_js();
+		$themes = wp_prepare_themes_for_js(
+			[
+				wp_get_theme( $active_theme_slug ),
+				wp_get_theme( $reader_theme_slug ),
```
Comment (westonruter) [inline tests/php/src/ReaderThemeLoaderTest.php:164]: Why were these needed to be added here explicitly I wonder?

File: tests/php/test-uninstall.php [hunk 4]
```diff
@@ -569,7 +569,17 @@ public function test_remove_plugin_data() {
...
 		require AMP__DIR__ . '/uninstall.php';
 		$this->flush_cache();
-		$this->assertEquals( $num_queries_before, $wpdb->num_queries );
+
+		if ( is_multisite() ) {
+			/*
+			* Multisite has 2 additional queries:
+			* 1. Get sites query.
+			* 2. Switch to blog query.
+			*/
```
Comment (westonruter) [inline tests/php/test-uninstall.php]:
```suggestion
			 * Multisite has 2 additional queries:
			 * 1. Get sites query.
			 * 2. Switch to blog query.
			 */
```

File: tests/php/test-amp-helper-functions.php [hunk 5]
```diff
@@ -2576,6 +2588,10 @@ public function data_amp_has_paired_endpoint() {
 	 * @param bool   $is_amp
 	 */
 	public function test_amp_has_paired_endpoint_go_to( $paired_url_structure, $suffix, $is_amp ) {
+		if ( is_multisite() ) {
+			switch_to_blog( self::factory()->blog->create() );
+		}
```
Comment (westonruter) [inline tests/php/test-amp-helper-functions.php]: Why is this needed?

File: .github/workflows/build-test-measure.yml [hunk 6]
```diff
@@ -5,7 +5,7 @@ on:
     branches:
       - develop
       # Include all release branches.
-      - '[0-9]+.[0-9]+'
+      - "[0-9]+.[0-9]+"
```
Comment (westonruter) [inline .github/workflows/build-test-measure.yml]: Use of double quotes instead of single quotes is just to improve convention?

**Review by westonruter — CHANGES_REQUESTED** [review]

File: .github/workflows/build-test-measure.yml (hunk 1 above)
Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml]: No, just used it for testing.

File: .github/workflows/build-test-measure.yml (hunk 6 above)
Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml]: Formatter did this. Will revert such changes.

File: tests/php/src/ReaderThemeLoaderTest.php (hunk 3 above)
Comment (thelovekesh) [reply tests/php/src/ReaderThemeLoaderTest.php:164]: On single-site installation, it returns all themes in an array but in the multisite environment, it only returns the current installed theme and a default theme.  Maybe because in respect to the current blog it can only fetch the theme that is installed?

File: tests/php/test-amp-helper-functions.php (hunk 5 above)
Comment (thelovekesh) [reply tests/php/test-amp-helper-functions.php]: Just to ensure the working with subdirectory installation i.e. on `/path1/some-post/amp`. `/path2/some-post/amp`. Or we can write another test to ensure this?

**Comment by westonruter** [comment]
> I believe the capability here should be `install_themes`. Thoughts?

@thelovekesh yes, I think you're right. The page requires `install_themes`: https://github.com/WordPress/wordpress-develop/blob/04f55a430e2ee5a88e9b5b5485b75a68fabd5a83/src/wp-admin/theme-install.php#L15-L17

File: tests/php/test-amp-helper-functions.php (hunk 5 above)
Comment (westonruter) [reply tests/php/test-amp-helper-functions.php]: Ah, ok. How about just adding a comment to indicate this is the reason for the blog switch?

File: tests/php/src/ReaderThemeLoaderTest.php (hunk 3 above)
Comment (westonruter) [reply tests/php/src/ReaderThemeLoaderTest.php:164]: Ah right. On a network, themes have to be network activated to be available for all sites.

File: tests/php/src/Admin/UserRESTEndpointExtensionTest.php [hunk 7]
```diff
@@ -68,7 +68,18 @@ public function test_update_review_panel_dismissed_for_template_mode() {
...
+		if ( is_multisite() ) {
+			$admin_user = self::factory()->user->create_and_get(
+				[
+					'role' => 'administrator',
+				]
+			);
+
+			grant_super_admin( $admin_user->ID );
+		}
+
```
Comment (westonruter) [inline tests/php/src/Admin/UserRESTEndpointExtensionTest.php]: This should also use `MockAdminUser`

Comment (thelovekesh) [reply tests/php/src/Admin/UserRESTEndpointExtensionTest.php]: @westonruter Please note the use of `create_and_get` and later this test also requires the WP user object.

Comment (westonruter) [reply tests/php/src/Admin/UserRESTEndpointExtensionTest.php]: You could make `MockAdminUser` return the object to use later.

File: tests/php/test-amp-helper-functions.php (hunk 5 above)
Comment (thelovekesh) [reply tests/php/test-amp-helper-functions.php]: I have added an inline comment.

File: tests/php/src/Helpers/MockAdminUser.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+	public function mock_admin_user() {
+		$admin_user = self::factory()->user->create_and_get( [ 'role' => 'administrator' ] );
+
+		if ( is_multisite() ) {
+			$admin_user = self::factory()->user->create_and_get(
+				[
+					'role' => 'administrator',
+				]
+			);
+
```
Comment (westonruter) [inline tests/php/src/Helpers/MockAdminUser.php]: This user object seems to be created as a duplicate.

```suggestion
```

Comment (thelovekesh) [reply tests/php/src/Helpers/MockAdminUser.php]: Ahh. yes.

**Review by westonruter — APPROVED** [review]
